substrate/client/network/sync/src/lib.rs

```diff
@@ -407,6 +407,122 @@ impl<B: BlockT> PeerSyncState<B> {
 	pub fn is_available(&self) -> bool {
 		matches!(self, Self::Available)
 	}
+
+	/// Checks if we can initiate a new download from the peer.
+	pub fn can_download(&self) -> bool {
+		self.is_available() || matches!(self, Self::DownloadingNew(_))
+	}
+}
+
+/// State about the download in progress.
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct PeerDownloadState<B: BlockT> {
+	/// peer.common_number when the download was started.
+	common_number: NumberFor<B>,
+
+	/// The range being download [start, end).
+	range: Range<NumberFor<B>>,
+
+	/// The blocks downloaded so far.
+	downloaded: Vec<BlockData<B>>,
+}
+
+impl<B: BlockT> PeerDownloadState<B> {
+	fn new(common_number: NumberFor<B>, range: Range<NumberFor<B>>) -> Self {
+		Self { common_number, range, downloaded: Vec::new() }
+	}
+
+	/// Handles the new blocks received from the peer.
+	/// Returns:
+	/// Ok(true): If done and no more requests need to be issued.
+	/// Ok(false): if more requests need to be issued.
+	/// Err(): On any failures.
+	fn handle_blocks(
+		&mut self,
+		who: &PeerId,
+		mut new_blocks: Vec<BlockData<B>>,
+		request: BlockRequest<B>,
+		collection: &mut BlockCollection<B>,
+	) -> Result<bool, BadPeer> {
+		collection.clear_peer_download(who);
+
+		// Validate the blocks.
+		let start_block = match validate_blocks::<B>(&new_blocks, who, Some(request)) {
+			Ok(Some(start_block)) => start_block,
+			Ok(None) => return Ok(true),
+			Err(err) => return Err(err),
+		};
+
+		// Add the new blocks to the downloaded list.
+		let range_len = (self.range.end - self.range.start).saturated_into::<u32>() as usize;
+		if (self.downloaded.len() + new_blocks.len()) > range_len {
+			return Err(BadPeer(*who, rep::BAD_RESPONSE))
+		}
+		let new_blocks_len = new_blocks.len();
+		new_blocks.append(&mut self.downloaded);
+		self.downloaded.append(&mut new_blocks);
+
+		if self.range.start == (self.common_number + One::one()) &&
+			self.downloaded.len() < range_len
+		{
+			// The download extends the common ancestor, but the response
+			// was partial. We want to issue further requests to fetch
+			// the incomplete part before importing.
+			trace!(
+				target: LOG_TARGET,
+				"Download state: incomplete download: \
+				common = {}, range = {:?}, new_blocks = {new_blocks_len}, downloaded = {}",
+				self.common_number, self.range, self.downloaded.len()
+			);
+			return Ok(false)
+		}
+
+		// Done, report the accumulated blocks to the collection.
+		let mut downloaded = Vec::new();
+		downloaded.append(&mut self.downloaded);
+		collection.insert(start_block, downloaded, *who);
```

Comment:
Same comment here, why is this `append()` necessary?

Reply:
This is similar, it moves the completed blocks out and passes it to the `BlockCollection`